---
title: "Exploring Alternative Splicing with Long Reads"
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    vertical_layout: fill
runtime: shiny
---

```{r global, include=FALSE}
library(flexdashboard)
library(data.table)
setDTthreads(8)
library(shiny)
projectDir="."
dataDir <- "/data/local/agosdsc/projects/mrg1_nanopore/flair_analyzed/pooled_hom_mrg1_v3/"
outDir=file.path(dataDir,"diffSplice_noShortReads/")
gtfFile_Ref="/fast/AG_Tursun/Base/Genome_Annotation/ce11/Ensembl/Caenorhabditis_elegans.WBcel235.97.gtf"
gtfFile <- file.path(dataDir,"collapse","all_samples_flair.isoforms.gtf")

# preppedDataFile <- file.path(dataDir,"preppedData.RDS")
preppedDataFile <- file.path(projectDir,"data","preppedData.RDS")

source(file.path(projectDir,"scripts","dashboard_funcs.R"))
if(!file.exists(preppedDataFile)) {
    
    tx_prod_full_file <- file.path(dataDir,"productivity","productivity.bed")
    diffSpliceFolder <- file.path(dataDir,"diffSplice_noShortReads")
    diffExpFolder <- file.path(dataDir,"diffExp_noShortReads")
    
    preppedData <- prepData(productivityBedFile = tx_prod_full_file, 
                         flairGtfFile = gtfFile, 
                         ensemblGtfFile = gtfFile_Ref, 
                         diffSpliceFolder = diffSpliceFolder,
                         diffExpFolder = diffExpFolder,
                         dataDir = dataDir
                         )
} else {
    preppedData <- readRDS(preppedDataFile)
}

# attach(preppedData)

suppressPackageStartupMessages({
  library(data.table)
  library(GenomicRanges)
  library(ggplot2)
  library(ggbio)
  library(reshape2)
  library(ggrepel)
  library(DESeq2)
  library(patchwork)
})

```


Sidebar {.sidebar data-width=250}
=====================================

#### Create a figure for the selected Gene:

```{r}
selectable <- with(preppedData, 
                  productivity[gene_exp[!is.na(gene_name),geneID], ,on = "gene_id"][
                    productivity %in% c("PRO","PTC"),unique(gene_id)])

selectInput("geneID", label = "Gene for Summary Figures:",
            choices = sort(preppedData$gene2symbol[selectable,gene_name,on = "gene_id"]), 
            selected = "che-13",
            # size = 5,
            multiple = FALSE,
            selectize = TRUE)
```

<!-- #### Download the figure: -->

<!-- ```{r} -->
<!-- # downloadObjUI(id = "download")) -->
<!-- downloadButton(outputId = "download", label = "Download the summary plot.") -->
<!-- #  -->
<!-- output$download <- -->
<!--   downloadHandler( -->
<!--     filename = function() { paste(geneSelection()$geneID, '.png', sep='') }, -->
<!--     content = function(file) { -->
<!--       # geneSelection()$p_summary -->

<!--       ggsave(filename = file, plot = plotSummary(), device = "png") -->
<!--     } -->
<!-- ) -->
<!-- #  -->
<!-- # # checkboxInput(inputId, label, value = FALSE, width = NULL) -->
<!-- ``` -->

#### Filter tables for Genes of Interest

```{r}
# Filter tables for Genes of Interest
selectInput("selectedGeneID", label = "Filter tables for Genes of Interest:",
            choices = sort(preppedData$gene2symbol[selectable,gene_name,on = "gene_id"]), 
            selected = NULL,
            multiple = TRUE,
            selectize = TRUE)

```

#### Filter to show only selected Productivity: 

```{r}
# Filter Productivity
checkboxGroupInput("filterProductivity", 
                   label = "Select Productivity", 
                   choices = list(
                     "productive transcript" = "PRO", 
                     "premature termination codon" = "PTC", 
                     "no start codon" = "NGO", 
                     "start but no stop codon" = "NST"
                     ),
                   selected = c("PRO","PTC","NST","NGO"))
```

### Filter to show only selected Alternative Splice Types: 

```{r}
# Filter ASEvents
checkboxGroupInput("filterASEvents", 
                   label = "Select Alternative Splice Types", 
                   choices = list(
                     "retained intron" = "ir",
                     "skipped exon" = "es",
                     "alt 3' splice site" = "alt3",
                     "alt 5' splice site" = "alt5"
                   ),
                   selected = c("ir","es","alt3","alt5")
                   )

```


Summarized Figures (Panel View)
=======================================================================

```{r define_plots, include=FALSE}

geneSelection <- reactive({

    ### filter geneID
    geneID <- input$geneID

    if(!geneID %in% preppedData$gene2symbol$gene_id) {
        geneID <- preppedData$gene2symbol[gene_name == geneID, gene_id]
    }

    geneName <- preppedData$gene2symbol[geneID, gene_name]
    
    return(list(geneID = geneID,geneName = geneName))
    
})

filteredProductivity <- reactive({
  
    ### filter productivity
    # filterProductivity <- c("PRO","PTC","NST","NGO")
    filterProductivity <- input$filterProductivity
    filterProductivity <- if(!is.null(filterProductivity)) {
        preppedData$productivity$productivity %in% filterProductivity
    } else {
      TRUE
    }
    filteredProductivity <- preppedData$productivity[filterProductivity,]
    
    return(filteredProductivity)
    
})

filteredAsEvents <- reactive({
  
    ### filter as types
    # filterProductivity <- c("PRO","PTC","NST","NGO")
    filterASEvents <- input$filterASEvents
    filterASEvents <- if(!is.null(filterASEvents)) {
        preppedData$asEvents$event %in% filterASEvents
    } else {
      TRUE
    }
    filteredAsEvents <- preppedData$asEvents[filterASEvents,]
    
    return(filteredAsEvents)
    
})
    

geneModel <- reactive({
  
    geneID <- geneSelection()$geneID
  
    ### generate figures
    txList <- extractTXFeatures(productivity = filteredProductivity(),
                                    gtf = preppedData$gtf,
                                    geneID = geneID)
    
    p_model <- plotGeneModel(txList = txList,
                             asEvents = filteredAsEvents(),
                             geneID = geneID,
                             tx2gene = preppedData$tx2gene)

    ## extract the exact order of tx labels
    txOrder <- with(layer_data(p_model, i = 6),label[order(y)])
    
    return(list(p_model = p_model, txOrder = txOrder))
    
})

boxplot <- reactive({
  
  geneID <- geneSelection()$geneID
  txOrder <- geneModel()$txOrder
  
  p_box <- plotTxExpression(norm.counts = preppedData$norm.counts,
                              geneID = geneID,
                              gene2symbol = preppedData$gene2symbol,
                              productivity = filteredProductivity(),
                              txOrder = txOrder,
                              meta = preppedData$meta,
                              labels = c("N2" = "N2","mrg1" = bquote("mrg-1"^{"-/-"})))
  
      return(list(p_box = p_box))
  
})


volcanoPlots <- reactive({
  
    geneID <- geneSelection()$geneID

    p_tx <- plotDTU(tx_exp_prod = preppedData$tx_exp_prod,
                    highlight_genes = geneID,
                    isGeneID = TRUE,
                    gene2symbol = preppedData$gene2symbol,
                    title = bquote("N2 vs mrg-1"^{"-/-"}))

    p_gene <- plotDGE(gene_exp = preppedData$gene_exp,
                      highlight_genes = geneID,
                      isGeneID=TRUE,
                      gene2symbol = preppedData$gene2symbol,
                      title = bquote("N2 vs mrg-1"^{"-/-"}))
    
    return(list(p_tx = p_tx,
                p_gene = p_gene))
    
})
  

summaryFigure <- reactive({
  
    geneID <- geneSelection()$geneID
  
    p_summary <- summaryPlot(geneID = geneID,
                   productivity = filteredProductivity(),
                   gtf = preppedData$gtf,
                   asEvents = filteredAsEvents(),
                   tx2gene = preppedData$tx2gene,
                   norm.counts = preppedData$norm.counts,
                   gene2symbol = preppedData$gene2symbol,
                   filterAsEvent = NULL,
                   isGeneID = TRUE,
                   meta = preppedData$meta,
                   tx_exp_prod = preppedData$tx_exp_prod,
                   gene_exp = preppedData$gene_exp
                   )
    
    return(list(p_summary = p_summary))
    
})




# summaryPlot(geneID = "adr-1",
#                    productivity = preppedData$productivity,
#                    gtf = preppedData$gtf,
#                    asEvents = preppedData$asEvents,
#                    tx2gene = preppedData$tx2gene,
#                    norm.counts = preppedData$norm.counts,
#                    gene2symbol = preppedData$gene2symbol,
#                    filterAsEvent = NULL,
#                    isGeneID = FALSE,
#                    meta = preppedData$meta,
#                    tx_exp_prod = preppedData$tx_exp_prod,
#                    gene_exp = preppedData$gene_exp
#                    )


```

Column {data-width=650}
-----------------------------------------------------------------------

### Differential Gene Expression

```{r}
# output$gene <- 
  renderPlot({
    volcanoPlots()$p_gene + coord_flip()
    })
# plotOutput("gene")
```


### Differential Transcript Expression

```{r}
# output$tx <- 
  renderPlot({
    volcanoPlots()$p_tx + coord_flip() + theme(legend.position = "bottom")
    })
# plotOutput("tx")
```

Column {data-width=350}
-----------------------------------------------------------------------

### Gene Model

```{r}
# output$model <- 
    renderPlot({
    geneModel()$p_model
    })

# plotOutput("model")
```


### Transcript Usage

```{r}
# output$box <- 
    renderPlot({
      boxplot()$p_box
      })
# plotOutput("box")
```

Summarized Figure (Aggregate View) 
=======================================================================

Column {data-width=500 data-height=800}
-----------------------------------------------------------------------

#### Figure

```{r ,fig.width=5, fig.height= 8}
# output$box <- 
    renderPlot({
      summaryFigure()$p_summary
      })
# plotOutput("box")
```

Result Tables {data-orientation=column vertical_layout=scroll}
=======================================================================

Column {data-height=200}
-----------------------------------------------------------------------

#### Alternative Splicing Results 

```{r}
DT::renderDataTable({
  
  selectedGeneID <- if(!is.null(input$selectedGeneID)) {
    preppedData$asEvents$gene_name %in% input$selectedGeneID
  } else {
    TRUE
  }
  
  DT::datatable(preppedData$asEvents[grepl("exclusion",feature_id) & selectedGeneID,
                                     c("gene_id","event","coordinate",
                                       "lr","pvalue","adj_pvalue","gene_name")][
                                         order(adj_pvalue,decreasing = FALSE)], 
                options = list(
                  pageLength = 10
                  ),
                rownames = FALSE)
})
```

#### Differential Gene Expression

```{r}
DT::renderDataTable({
  
  selectedGeneID <- if(!is.null(input$selectedGeneID)) {
    preppedData$gene_exp$gene_name %in% input$selectedGeneID
  } else {
    TRUE
  }
  
  DT::datatable(preppedData$gene_exp[selectedGeneID][
                                         order(padj,decreasing = FALSE)], 
                options = list(
                  pageLength = 10
                  ),rownames = FALSE)
})
```

#### Differential Transcript Expression

```{r}
DT::renderDataTable({
  
  selectedGeneID <- if(!is.null(input$selectedGeneID)) {
    preppedData$tx_exp_prod$gene_name %in% input$selectedGeneID
  } else {
    TRUE
  }
  
  DT::datatable(preppedData$tx_exp_prod[selectedGeneID,
                                        c("gene_id","feature_id","log2FoldChange",
                                          "lr","pvalue","padj","gene_name",
                                          "productivity")][
                                         order(padj,decreasing = FALSE)], 
                options = list(
                  pageLength = 10
                  ),rownames = FALSE)
})
```
